src/batou_ext/oci.py

```diff
@@ -158,69 +161,80 @@ def verify(self):
                 """
             )
         )
+
+        # If the container is not running the image we expect, we need to restart it.
+        # This will also ensure the container is up-to-date locally since the
+        # container's nix service is set to always pull
         if local_image_id != container_image_id:
-            # If the container is not running the image we expect, we need to
-            # restart it. If its the same, we need to dig further
             error = (
                 "Container is running different image. "
                 "({container_image_id} vs. {local_image_id})"
             )
-            local_digest, stderr = self.cmd(
-                dedent(
-                    """\
-                docker image inspect {{component.image}}:{{component.version}} \
-                    | jq -r 'first | .RepoDigests | first | split("@") | last' \
-                    || echo image not available locally
-                    """
-                )
+            self.log(
+                "Container is running an older version than is locally available and needs to be restarted"
             )
+            raise UpdateNeeded()
 
-            image_ident = f"{self.image}:{self.version}@{local_digest}"
-            try:
-                valid = self._remote_manifest_cache[image_ident]
-                error = "(cached)"
-            except KeyError:
-                if self.registry_address:
-                    logintxt, _ = self.cmd(
-                        self.expand(
-                            dedent(
-                                """\
-                docker login \\
-                    {%- if component.registry_user and component.registry_password %}
-                    -u {{component.registry_user}} \\
-                    -p {{component.registry_password}} \\
-                    {%- endif %}
-                    {{component.registry_address}}
+        # If the container is running the newest locally available image then
+        # check if there is a newer version available remotely
+
+        # query the local digest to compare against upstream
+        local_digest, stderr = self.cmd(
+            dedent(
+                """\
+            docker image inspect {{component.image}}:{{component.version}} \
+                | jq -r 'first | .RepoDigests | first | split("@") | last' \
+                || echo image not available locally
                 """
-                            )
-                        )
-                    )
+            )
+        )
+
+        image_ident = f"{self.image}:{self.version}@{local_digest}"
 
-                try:
-                    stdout, stderr = self.cmd(
-                        f"docker manifest inspect {image_ident}"
+        # test whether the ident has been checked already
+        if image_ident in self._remote_manifest_cache:
+            if self._remote_manifest_cache[image_ident]:
```

Comment:
there's essentially 3 possible situations here:

- not in dict -> needs remote lookup
- in dict and true -> does not need an update
- in dict and false -> requires an update

with a .get() case 1 and 3 would be equivalent because the truthiness of both None and False is False